1/dn01/scripts/demo.jl

```diff
@@ -0,0 +1,172 @@
+Argument `sprem` je vektor vozlišč grafa, ki nimajo določenih koordinat.
+Indeksi v matriki `A` ustrezajo vozliščem v istem vrstnem redu,
+kot nastopajo v argumentu `sprem`.
+"""
+function matrika(G::AbstractGraph, sprem)
+    # preslikava med vozlišči in indeksi v matriki
+    v_to_i = Dict([sprem[i] => i for i in eachindex(sprem)])
+    m = length(sprem)
+    A = RedkaMatrika(zeros(m, 1), zeros(m, 1))
+    for i = 1:m
+        vertex = sprem[i]
+        sosedi = neighbors(G, vertex)
+        for vertex2 in sosedi
+            if haskey(v_to_i, vertex2)
+                j = v_to_i[vertex2]
+                A[i, j] = 1.0
+            end
+        end
+        A[i, i] = float(-length(sosedi))
+    end
+    return A
+end
+
+
+"""
+G = krožna_lestev(n)
+Ustvari graf krožna lestev z `2n` točkami.
+"""
+function krožna_lestev(n)
+    G = SimpleGraph(2 * n)
+    # prvi cikel
+    for i = 1:n-1
+        add_edge!(G, i, i + 1)
+    end
+    add_edge!(G, 1, n)
+    # drugi cikel
+    for i = n+1:2n-1
+        add_edge!(G, i, i + 1)
+    end
+    add_edge!(G, n + 1, 2n)
+    # povezave med obema cikloma
+    for i = 1:n
+        add_edge!(G, i, i + n)
+    end
+    return G
+end
+
+using Logging
+"""
+x = cg(A, b; atol=1e-10)
+Metoda konjugiranih gradientov za reševanje sistema enačb `Ax = b`
+s pozitivno definitno matriko `A`. Argument `A` ni nujno matrika, lahko je
+tudi drugega tipa, če ima implementirano množenje z vektorjem `b`.
+Metoda ne preverja, ali je argument `A` pozitivno definiten.
+"""
+function cg(A, b; atol=1e-8)
+    # za začetni približek vzamemo kar desne strani
+    x = copy(b)
+    r = b - A * b
+    p = r
+    res0 = sum(r .* r)
+    for i = 1:length(b)
+        Ap = A * p
+        alpha = res0 / sum(p .* Ap)
+        x = x + alpha * p
+        r = r - alpha * Ap
+        res1 = sum(r .* r)
+        if sqrt(res1) < atol
+            @info "Metoda KG konvergira po $i korakih."
+            break
+        end
+        p = r + (res1 / res0) * p
+        res0 = res1
+    end
+    return x
+end
+
+
+#Vložitev 
+function vlozitev()
+    n = 10
+    G = krožna_lestev(n)
+    sprem = setdiff(vertices(G), 1:5) #[1,2,3....50]
+    A = matrika(G, sprem)
+
+    b = desne_strani(G, sprem, rand(5))
+
+    x,i = sor(A, b, rand(length(b)), 0.5)
+
+    return x
+end
+
+
+using Plots
+function najdi_najboljsi_omega(rm::RedkaMatrika, b, x0) 
+    segments = 20
+    omegas = LinRange(0, 2, segments)
+
+    is = []
+    ws = []
+    best_i = 1000
+    best_w = -1
+    for k = 1:segments
+        w = omegas[k]
+
+        try
+            x, i = sor(rm, b, x0, w)
+            push!(is, i)
+            push!(ws, w)
+
+            if i < best_i
+                best_i = i
+                best_w = w
+            end
+        catch e
+            push!(is, 1000)
+            push!(ws, w)
+        end
+   
+    end
+
+    plot(ws, is, xlabel="Omege", ylabel="Št. iteracij", title="Število iteracij do konvergence glede na omego")
```

Comment:
Risanje grafa znotraj funkcije, ki računa optimalni omega, ni najboljša strategija. Funkciji doda [stranski učinek](https://en.wikipedia.org/wiki/Side_effect_(computer_science), kar naredi funkcijo manj uporabno za ponovno uporabo. Bolje je risanje grafa in računanje optimalnega omega ločiti v dve funkciji. Več o [čistih funkcijah](https://docs.scala-lang.org/scala3/book/fp-pure-functions.html).